File: examples/python/m1-unified-protocol/openai_responses.py
"""OpenAI Responses API 与统一协议之间的编解码。

上层只看见 Message / ModelRequest / ModelResponse；本模块负责 Provider JSON。
"""

from __future__ import annotations

import json
from typing import Any, Mapping

from protocol import (
    ApiError,
    ContentBlock,
    DecodeError,
    EncodeError,
    Message,
    ModelRequest,
    ModelResponse,
    Role,
    TextBlock,
    ToolResultBlock,
    ToolUseBlock,
)


def encode_request(request: ModelRequest) -> dict[str, Any]:
    """将统一请求编码为 Responses API JSON Body。"""

    items: list[dict[str, Any]] = []
    if request.system and request.system.strip():
        items.append(
            {
                "type": "message",
                "role": "system",
                "content": [{"type": "input_text", "text": request.system}],
            }
        )

    for message in request.messages:
        items.extend(_encode_message(message))

    if not items:
        raise EncodeError("encoded request input must not be empty")

    payload: dict[str, Any] = {
        "model": request.model,
        "input": items,
    }
    if request.tools:
        payload["tools"] = [
            {
                "type": "function",
                "name": tool.name,
                "description": tool.description,
                "parameters": tool.input_schema,
                "strict": tool.strict,
            }
            for tool in request.tools
        ]
    return payload


def decode_response(body: Mapping[str, Any]) -> ModelResponse:
    """将 Responses API JSON 解码为统一响应。"""

    output = body.get("output")
    if not isinstance(output, list):
        raise DecodeError("API response did not contain an output list")

    blocks: list[ContentBlock] = []
    for item in output:
        if not isinstance(item, dict):
            continue
        item_type = item.get("type")
        if item_type == "message":
            blocks.extend(_decode_message_item(item))
        elif item_type == "function_call":
            blocks.append(_decode_function_call(item))
        # reasoning 及其他未知项目直接跳过，不当作协议成功内容。

    if not blocks:
        raise DecodeError("API response contained no decodable content blocks")

    response_id = body.get("id")
    model = body.get("model")
    status = body.get("status")
    if status is not None and not isinstance(status, str):
        status = None

    return ModelResponse(
        id=response_id if isinstance(response_id, str) else None,
        model=model if isinstance(model, str) else None,
        message=Message(role=Role.ASSISTANT, content=tuple(blocks)),
        status=status,
    )


def decode_response_json(text: str) -> ModelResponse:
    """解析 JSON 字符串并解码为统一响应。"""

    try:
        body = json.loads(text)
    except json.JSONDecodeError as error:
        raise DecodeError("API returned invalid JSON") from error

    if not isinstance(body, dict):
        raise DecodeError("API response must be a JSON object")
    return decode_response(body)


def parse_http_response(status_code: int, body: str) -> ModelResponse:
    """检查 HTTP 状态后解码响应；错误不回显正文。"""

    if not 200 <= status_code < 300:
        raise ApiError(f"API request failed with HTTP {status_code}")
    return decode_response_json(body)


def _encode_message(message: Message) -> list[dict[str, Any]]:
    """把一条统一消息编码成一个或多个 Responses input 项目。"""

    if message.role == Role.TOOL:
        return [_encode_tool_result_item(block) for block in message.content]

    if message.role == Role.ASSISTANT:
        return _encode_assistant_items(message)

    role = message.role.value
    content_items: list[dict[str, Any]] = []
    for block in message.content:
        if isinstance(block, TextBlock):
            content_items.append({"type": "input_text", "text": block.text})
        elif isinstance(block, ToolResultBlock):
            raise EncodeError("tool_result blocks require role=tool")
        elif isinstance(block, ToolUseBlock):
            raise EncodeError("tool_use blocks require role=assistant")
        else:
            raise EncodeError(f"unsupported content block: {type(block)!r}")

    if not content_items:
        raise EncodeError("message produced no provider content items")

    return [{"type": "message", "role": role, "content": content_items}]


def _encode_assistant_items(message: Message) -> list[dict[str, Any]]:
    """assistant 消息可能同时包含文本和 tool_use，需拆成多个 output 形态 input。"""

    items: list[dict[str, Any]] = []
    text_parts: list[dict[str, Any]] = []

    def flush_text() -> None:
        nonlocal text_parts
        if text_parts:
            items.append(
                {
                    "type": "message",
                    "role": "assistant",
                    "content": text_parts,
                }
            )
            text_parts = []

    for block in message.content:
        if isinstance(block, TextBlock):
            text_parts.append({"type": "output_text", "text": block.text})
        elif isinstance(block, ToolUseBlock):
            flush_text()
            items.append(
                {
                    "type": "function_call",
                    "call_id": block.id,
                    "name": block.name,
                    "arguments": json.dumps(block.input, ensure_ascii=False),
                }
            )
        elif isinstance(block, ToolResultBlock):
            raise EncodeError("assistant message cannot contain tool_result")
        else:
            raise EncodeError(f"unsupported content block: {type(block)!r}")

    flush_text()
    if not items:
        raise EncodeError("assistant message produced no provider items")
    return items


def _encode_tool_result_item(block: ContentBlock) -> dict[str, Any]:
    """将 tool_result 编码为 function_call_output。"""

    if not isinstance(block, ToolResultBlock):
        raise EncodeError("role=tool messages may only contain tool_result blocks")

    return {
        "type": "function_call_output",
        "call_id": block.tool_use_id,
        "output": block.content,
    }


def _decode_message_item(item: Mapping[str, Any]) -> list[ContentBlock]:
    """从 message 输出项中提取文本块。"""

    content = item.get("content")
    if not isinstance(content, list):
        return []

    blocks: list[ContentBlock] = []
    for content_item in content:
        if not isinstance(content_item, dict):
            continue
        content_type = content_item.get("type")
        if content_type in {"output_text", "text", "input_text"}:
            text = content_item.get("text")
            if isinstance(text, str) and text.strip():
                blocks.append(TextBlock(text=text))
        elif content_type == "refusal":
            refusal = content_item.get("refusal")
            if isinstance(refusal, str) and refusal.strip():
                blocks.append(TextBlock(text=refusal))
    return blocks


def _decode_function_call(item: Mapping[str, Any]) -> ToolUseBlock:
    """把 Responses function_call 项目解码为 ToolUseBlock。"""

    call_id = item.get("call_id") or item.get("id")
    name = item.get("name")
    if not isinstance(call_id, str) or not call_id.strip():
        raise DecodeError("function_call is missing call_id")
    if not isinstance(name, str) or not name.strip():
        raise DecodeError("function_call is missing name")

    arguments = item.get("arguments", "{}")
    parsed = _parse_arguments(arguments)
    return ToolUseBlock(id=call_id, name=name, input=parsed)


def _parse_arguments(arguments: Any) -> dict[str, Any]:
    """arguments 可能是 JSON 字符串或对象；统一成 dict。"""

    if isinstance(arguments, dict):
        return dict(arguments)
    if arguments is None:
        return {}
    if not isinstance(arguments, str):
        raise DecodeError("function_call arguments must be an object or JSON string")

    text = arguments.strip()
    if not text:
        return {}

    try:
        value = json.loads(text)
    except json.JSONDecodeError as error:
        raise DecodeError("function_call arguments are not valid JSON") from error

    if value is None:
        return {}
    if not isinstance(value, dict):
        raise DecodeError("function_call arguments JSON must be an object")
    return value

File: examples/rust/m1-unified-protocol/src/openai_responses.rs
//! OpenAI Responses API 与统一协议之间的编解码。

use crate::protocol::{
    ContentBlock, Message, ModelRequest, ModelResponse, ProtocolError, Role, TextBlock,
    ToolResultBlock, ToolUseBlock,
};
use serde_json::{Value, json};

/// 将统一请求编码为 Responses API JSON Body。
pub fn encode_request(request: &ModelRequest) -> Result<Value, ProtocolError> {
    let mut items = Vec::new();

    if let Some(system) = request
        .system
        .as_ref()
        .map(|text| text.trim())
        .filter(|text| !text.is_empty())
    {
        items.push(json!({
            "type": "message",
            "role": "system",
            "content": [{"type": "input_text", "text": system}],
        }));
    }

    for message in &request.messages {
        items.extend(encode_message(message)?);
    }

    if items.is_empty() {
        return Err(ProtocolError::Encode(
            "encoded request input must not be empty".to_owned(),
        ));
    }

    let mut payload = json!({
        "model": request.model,
        "input": items,
    });
    if !request.tools.is_empty() {
        payload["tools"] = Value::Array(
            request
                .tools
                .iter()
                .map(|tool| {
                    json!({
                        "type": "function",
                        "name": tool.name,
                        "description": tool.description,
                        "parameters": tool.input_schema,
                        "strict": tool.strict,
                    })
                })
                .collect(),
        );
    }
    Ok(payload)
}

/// 将 Responses API JSON 解码为统一响应。
pub fn decode_response(body: &Value) -> Result<ModelResponse, ProtocolError> {
    let output = body
        .get("output")
        .and_then(Value::as_array)
        .ok_or_else(|| {
            ProtocolError::Decode("API response did not contain an output list".into())
        })?;

    let mut blocks = Vec::new();
    for item in output {
        let Some(item_type) = item.get("type").and_then(Value::as_str) else {
            continue;
        };
        match item_type {
            "message" => blocks.extend(decode_message_item(item)),
            "function_call" => {
                blocks.push(ContentBlock::ToolUse(decode_function_call(item)?));
            }
            _ => {}
        }
    }

    if blocks.is_empty() {
        return Err(ProtocolError::Decode(
            "API response contained no decodable content blocks".to_owned(),
        ));
    }

    let id = body.get("id").and_then(Value::as_str).map(str::to_owned);
    let model = body.get("model").and_then(Value::as_str).map(str::to_owned);
    let status = body
        .get("status")
        .and_then(Value::as_str)
        .map(str::to_owned);

    Ok(ModelResponse {
        id,
        model,
        message: Message::try_new(Role::Assistant, blocks)?,
        status,
    })
}

/// 解析 JSON 字符串并解码为统一响应。
pub fn decode_response_json(text: &str) -> Result<ModelResponse, ProtocolError> {
    let body: Value = serde_json::from_str(text)
        .map_err(|_| ProtocolError::Decode("API returned invalid JSON".to_owned()))?;
    if !body.is_object() {
        return Err(ProtocolError::Decode(
            "API response must be a JSON object".to_owned(),
        ));
    }
    decode_response(&body)
}

/// 检查 HTTP 状态后解码响应；错误不回显正文。
pub fn parse_http_response(status_code: u16, body: &str) -> Result<ModelResponse, ProtocolError> {
    if !(200..300).contains(&status_code) {
        return Err(ProtocolError::Api(format!(
            "API request failed with HTTP {status_code}"
        )));
    }
    decode_response_json(body)
}

fn encode_message(message: &Message) -> Result<Vec<Value>, ProtocolError> {
    match message.role {
        Role::Tool => message
            .content
            .iter()
            .map(encode_tool_result_item)
            .collect(),
        Role::Assistant => encode_assistant_items(message),
        Role::System | Role::User => {
            let mut content_items = Vec::new();
            for block in &message.content {
                match block {
                    ContentBlock::Text(text) => {
                        content_items.push(json!({
                            "type": "input_text",
                            "text": text.text,
                        }));
                    }
                    ContentBlock::ToolResult(_) => {
                        return Err(ProtocolError::Encode(
                            "tool_result blocks require role=tool".to_owned(),
                        ));
                    }
                    ContentBlock::ToolUse(_) => {
                        return Err(ProtocolError::Encode(
                            "tool_use blocks require role=assistant".to_owned(),
                        ));
                    }
                }
            }
            if content_items.is_empty() {
                return Err(ProtocolError::Encode(
                    "message produced no provider content items".to_owned(),
                ));
            }
            Ok(vec![json!({
                "type": "message",
                "role": message.role.as_str(),
                "content": content_items,
            })])
        }
    }
}

fn encode_assistant_items(message: &Message) -> Result<Vec<Value>, ProtocolError> {
    let mut items = Vec::new();
    let mut text_parts = Vec::new();

    let flush_text = |text_parts: &mut Vec<Value>, items: &mut Vec<Value>| {
        if !text_parts.is_empty() {
            items.push(json!({
                "type": "message",
                "role": "assistant",
                "content": text_parts.clone(),
            }));
            text_parts.clear();
        }
    };

    for block in &message.content {
        match block {
            ContentBlock::Text(text) => {
                text_parts.push(json!({
                    "type": "output_text",
                    "text": text.text,
                }));
            }
            ContentBlock::ToolUse(tool_use) => {
                flush_text(&mut text_parts, &mut items);
                let arguments = serde_json::to_string(&tool_use.input).map_err(|_| {
                    ProtocolError::Encode("tool_use input is not JSON-serializable".to_owned())
                })?;
                items.push(json!({
                    "type": "function_call",
                    "call_id": tool_use.id,
                    "name": tool_use.name,
                    "arguments": arguments,
                }));
            }
            ContentBlock::ToolResult(_) => {
                return Err(ProtocolError::Encode(
                    "assistant message cannot contain tool_result".to_owned(),
                ));
            }
        }
    }
    flush_text(&mut text_parts, &mut items);

    if items.is_empty() {
        return Err(ProtocolError::Encode(
            "assistant message produced no provider items".to_owned(),
        ));
    }
    Ok(items)
}

fn encode_tool_result_item(block: &ContentBlock) -> Result<Value, ProtocolError> {
    let ContentBlock::ToolResult(result) = block else {
        return Err(ProtocolError::Encode(
            "role=tool messages may only contain tool_result blocks".to_owned(),
        ));
    };
    Ok(json!({
        "type": "function_call_output",
        "call_id": result.tool_use_id,
        "output": result.content,
    }))
}

fn decode_message_item(item: &Value) -> Vec<ContentBlock> {
    let Some(content) = item.get("content").and_then(Value::as_array) else {
        return Vec::new();
    };

    let mut blocks = Vec::new();
    for content_item in content {
        let Some(content_type) = content_item.get("type").and_then(Value::as_str) else {
            continue;
        };
        match content_type {
            "output_text" | "text" | "input_text" => {
                if let Some(text) = content_item
                    .get("text")
                    .and_then(Value::as_str)
                    .filter(|text| !text.trim().is_empty())
                {
                    blocks.push(ContentBlock::Text(TextBlock {
                        text: text.to_owned(),
                    }));
                }
            }
            "refusal" => {
                if let Some(text) = content_item
                    .get("refusal")
                    .and_then(Value::as_str)
                    .filter(|text| !text.trim().is_empty())
                {
                    blocks.push(ContentBlock::Text(TextBlock {
                        text: text.to_owned(),
                    }));
                }
            }
            _ => {}
        }
    }
    blocks
}

fn decode_function_call(item: &Value) -> Result<ToolUseBlock, ProtocolError> {
    let call_id = item
        .get("call_id")
        .or_else(|| item.get("id"))
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .ok_or_else(|| ProtocolError::Decode("function_call is missing call_id".into()))?;
    let name = item
        .get("name")
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .ok_or_else(|| ProtocolError::Decode("function_call is missing name".into()))?;
    let input = parse_arguments(item.get("arguments").unwrap_or(&Value::Null))?;
    Ok(ToolUseBlock {
        id: call_id.to_owned(),
        name: name.to_owned(),
        input,
    })
}

fn parse_arguments(arguments: &Value) -> Result<Value, ProtocolError> {
    match arguments {
        Value::Object(_) => Ok(arguments.clone()),
        Value::Null => Ok(json!({})),
        Value::String(text) => {
            let trimmed = text.trim();
            if trimmed.is_empty() {
                return Ok(json!({}));
            }
            let value: Value = serde_json::from_str(trimmed).map_err(|_| {
                ProtocolError::Decode("function_call arguments are not valid JSON".to_owned())
            })?;
            if value.is_null() {
                return Ok(json!({}));
            }
            if !value.is_object() {
                return Err(ProtocolError::Decode(
                    "function_call arguments JSON must be an object".to_owned(),
                ));
            }
            Ok(value)
        }
        _ => Err(ProtocolError::Decode(
            "function_call arguments must be an object or JSON string".to_owned(),
        )),
    }
}

/// 构造只含一个 tool_result 的 tool 角色消息。
pub fn tool_result_message(
    tool_use_id: impl Into<String>,
    content: impl Into<String>,
) -> Result<Message, ProtocolError> {
    Message::try_new(
        Role::Tool,
        vec![ContentBlock::ToolResult(ToolResultBlock {
            tool_use_id: tool_use_id.into(),
            content: content.into(),
            is_error: false,
        })],
    )
}
